src/backend/cpu/sift.cpp

```diff
@@ -0,0 +1,1046 @@
+    for (int y = ImgBorder; y < idims[1]-ImgBorder; y++) {
+        for (int x = ImgBorder; x < idims[0]-ImgBorder; x++) {
+            float p = center_ptr[y*idims[0] + x];
+
+            // Find extrema
+            if (abs((float)p) > threshold &&
+                ((p > 0                                && p > center_ptr[(y-1)*idims[0] + x-1] && p > center_ptr[(y-1)*idims[0] + x]   &&
+                  p > center_ptr[(y-1)*idims[0] + x+1] && p > center_ptr[y*idims[0] + (x-1)]   && p > center_ptr[y*idims[0] + x+1]     &&
+                  p > center_ptr[(y+1)*idims[0] + x-1] && p > center_ptr[(y+1)*idims[0] + x]   && p > center_ptr[(y+1)*idims[0] + x+1] &&
+                  p > prev_ptr[(y-1)*idims[0] + x-1]   && p > prev_ptr[(y-1)*idims[0] + x]     && p > prev_ptr[(y-1)*idims[0] + x+1]   &&
+                  p > prev_ptr[y*idims[0] + x-1]       && p > prev_ptr[y*idims[0] + x]         && p > prev_ptr[y*idims[0] + x+1]       &&
+                  p > prev_ptr[(y+1)*idims[0] + x-1]   && p > prev_ptr[(y+1)*idims[0] + x]     && p > prev_ptr[(y+1)*idims[0] + x+1]   &&
+                  p > next_ptr[(y-1)*idims[0] + x-1]   && p > next_ptr[(y-1)*idims[0] + x]     && p > next_ptr[(y-1)*idims[0] + x+1]   &&
+                  p > next_ptr[y*idims[0] + x-1]       && p > next_ptr[y*idims[0] + x]         && p > next_ptr[y*idims[0] + x+1]       &&
+                  p > next_ptr[(y+1)*idims[0] + x-1]   && p > next_ptr[(y+1)*idims[0] + x]     && p > next_ptr[(y+1)*idims[0] + x+1])  ||
+                 (p < 0                                && p < center_ptr[(y-1)*idims[0] + x-1] && p < center_ptr[(y-1)*idims[0] + x]   &&
+                  p < center_ptr[(y-1)*idims[0] + x+1] && p < center_ptr[y*idims[0] + (x-1)]   && p < center_ptr[y*idims[0] + x+1]     &&
+                  p < center_ptr[(y+1)*idims[0] + x-1] && p < center_ptr[(y+1)*idims[0] + x]   && p < center_ptr[(y+1)*idims[0] + x+1] &&
+                  p < prev_ptr[(y-1)*idims[0] + x-1]   && p < prev_ptr[(y-1)*idims[0] + x]     && p < prev_ptr[(y-1)*idims[0] + x+1]   &&
+                  p < prev_ptr[y*idims[0] + x-1]       && p < prev_ptr[y*idims[0] + x]         && p < prev_ptr[y*idims[0] + x+1]       &&
+                  p < prev_ptr[(y+1)*idims[0] + x-1]   && p < prev_ptr[(y+1)*idims[0] + x]     && p < prev_ptr[(y+1)*idims[0] + x+1]   &&
+                  p < next_ptr[(y-1)*idims[0] + x-1]   && p < next_ptr[(y-1)*idims[0] + x]     && p < next_ptr[(y-1)*idims[0] + x+1]   &&
+                  p < next_ptr[y*idims[0] + x-1]       && p < next_ptr[y*idims[0] + x]         && p < next_ptr[y*idims[0] + x+1]       &&
+                  p < next_ptr[(y+1)*idims[0] + x-1]   && p < next_ptr[(y+1)*idims[0] + x]     && p < next_ptr[(y+1)*idims[0] + x+1]))) {
+
+                if (*counter < max_feat)
+                {
+                    x_out[*counter] = (float)y;
+                    y_out[*counter] = (float)x;
+                    layer_out[*counter] = layer;
+                    (*counter)++;
```

Comment:
It could be a return value, but I used it as is to make it more similar to CUDA and OpenCL implementations, where the counter must be a device pointer.